src/internal/storage/fileset/prefetcher.go

```diff
@@ -0,0 +1,140 @@
+package fileset
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/chunk"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset/index"
+	"golang.org/x/sync/semaphore"
+)
+
+// prefetcher prefetches chunks for the files emitted by the wrapped file set.
+// The prefetching applies at chunk boundaries rather than file boundaries, so
+// it handles the prefetching of chunks that contain the content of many small
+// files as well as chunks that contain the content of a larger file. This is
+// handled by buffering file metadata up to chunk boundaries, then kicking off
+// concurrent tasks to prefetch the chunks and emit the files associated with
+// the chunks. The prefetching of files with more chunks than the prefetch
+// limit will not be handled by this abstraction. In this case, the prefetching
+// will be handled when the content of the file is needed.
+type prefetcher struct {
+	FileSet
+	storage *Storage
+}
+
+func NewPrefetcher(storage *Storage, fileSet FileSet) FileSet {
+	return &prefetcher{
+		FileSet: fileSet,
+		storage: storage,
+	}
+}
+
+func (p *prefetcher) Iterate(ctx context.Context, cb func(File) error, opts ...index.Option) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	taskChain := chunk.NewTaskChain(ctx, semaphore.NewWeighted(int64(p.storage.prefetchLimit)))
+	fetchChunk := func(ref *chunk.DataRef, files []File) error {
+		return taskChain.CreateTask(func(ctx context.Context) (func() error, error) {
+			if ref != nil {
+				if err := p.storage.ChunkStorage().PrefetchData(ctx, ref); err != nil {
+					return nil, err
+				}
+			}
+			return func() error {
+				for _, f := range files {
+					if err := cb(f); err != nil {
+						return err
+					}
+				}
+				return nil
+			}, nil
+		})
+	}
+	waitFile := func(f File) error {
+		done := make(chan struct{})
+		if err := taskChain.CreateTask(func(_ context.Context) (func() error, error) {
+			return func() error {
+				defer close(done)
+				return cb(f)
+			}, nil
+		}); err != nil {
+			return err
+		}
+		select {
+		case <-done:
+			return nil
+		case <-ctx.Done():
+			return errors.EnsureStack(ctx.Err())
+		}
+	}
+	var ref *chunk.DataRef
+	var files []File
+	maxFilesBuf := p.storage.shardCountThreshold / int64(p.storage.prefetchLimit)
+	if err := p.FileSet.Iterate(ctx, func(f File) error {
+		// Emit the files if a large number are buffered.
+		if int64(len(files)) >= maxFilesBuf {
+			if err := fetchChunk(ref, files); err != nil {
+				return err
+			}
+			ref = nil
+			files = nil
+		}
+		dataRefs := f.Index().File.DataRefs
+		// Handle empty files.
+		// It does not matter which chunk they get emitted with,
+		// just that they are emitted in the correct order.
+		if len(dataRefs) == 0 {
+			files = append(files, f)
+			return nil
+		}
+		// Handle first chunk.
```

Comment:
This comment might be a little misleading. If we have a file that spans multiple chunks, once we prefetch all the chunks and emit the file on the last chunk, `ref` and `files` will both be `nil`. In that case this code will trigger again.
At the same time I'm not sure how to best phrase what it is actually doing.

Reply:
I refactored the code a bit so we don't need to treat this as a special case.